frontend/src/metabase/visualizations/visualizations/Heading/Heading.tsx

```diff
@@ -0,0 +1,84 @@
+import { useMemo, MouseEvent } from "react";
+
+import cx from "classnames";
+import { t } from "ttag";
+
+import { useFocus } from "metabase/hooks/use-focus";
+import type {
+  BaseDashboardOrderedCard,
+  VisualizationSettings,
+} from "metabase-types/api";
+
+import {
+  InputContainer,
+  HeadingContent,
+  HeadingContainer,
+  TextInput,
+} from "./Heading.styled";
+
+interface HeadingProps {
+  isEditing: boolean;
+  onUpdateVisualizationSettings: ({ text }: { text: string }) => void;
+  dashcard: BaseDashboardOrderedCard;
+  settings: VisualizationSettings;
+}
+
+export function Heading({
+  settings,
+  isEditing,
+  onUpdateVisualizationSettings,
+  dashcard,
+}: HeadingProps) {
+  const justAdded = useMemo(() => dashcard?.justAdded || false, [dashcard]);
+
+  const { isFocused, toggleFocusOn, toggleFocusOff } = useFocus(justAdded);
+  const isPreviewing = !isFocused;
+
+  const handleTextChange = (text: string) =>
+    onUpdateVisualizationSettings({ text });
+  const preventDragging = (e: MouseEvent<HTMLInputElement>) =>
+    e.stopPropagation();
+
+  const content = settings.text;
+  const hasNoContent = !content;
+  const placeholder = t`Heading`;
+
+  if (isEditing) {
+    return (
+      <InputContainer
+        className={cx("text-edit-container", {
+          "has-no-content": hasNoContent,
+        })}
+        hasNoContent={hasNoContent}
+        isPreviewing={isPreviewing}
+        onClick={toggleFocusOn}
+      >
+        {isPreviewing ? (
+          <HeadingContent
+            hasNoContent={hasNoContent}
+            isEditing={isEditing}
+            onMouseDown={preventDragging}
+          >
+            {content || placeholder}
```

Comment:
are we sure we'll never get a numeric zero here? If not, maybe try using `isEmpty` ?

https://github.com/metabase/metabase/blob/ffc2db4932d6affc794eae85c6fddf49acd78252/frontend/src/metabase/lib/validate.js#L6-L7


```suggestion
            {isEmpty(content) ? content : placeholder}
```